Add QDP data file handling

matteobachetti opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

This pull request is to address Issue #8382, adding QDP data handling.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #8382

pep8speaks commented:
Hello @matteobachetti :wave:! Thanks for opening this pull request, we are very grateful for your contribution! I'm a friendly :robot: that checks for style issues in this pull request, since this project follows the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guidelines. I've listed some small issues I found below, but please don't hesitate to ask if any of them are unclear!




* In the file [`astropy/io/misc/qdp.py`](https://github.com/astropy/astropy/blob/16fb04a8f054f2bf885fdc46b0c53eff6d5055ae/astropy/io/misc/qdp.py):

> [Line 74:5](https://github.com/astropy/astropy/blob/16fb04a8f054f2bf885fdc46b0c53eff6d5055ae/astropy/io/misc/qdp.py#L74): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

embray commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+    if input_colnames is None or len(input_colnames) == ncol:
+        colnames = interpret_err_lines(
+            err_specs, ncol, input_colnames=input_colnames
+        )
```
I don't really understand what this is doing.  Is it initializing the `err_specs` data structure?

matteobachetti replied:
It is reading err_specs and understanding if some columns represent the uncertainty of others. Main column are named col1, col2, ... or whatever the user gives (e.g. MJD, Rate). Error column are named col0_err (or MJD_err) for symmetric error, col0_perr and col0_nerr for positive/negative error

matteobachetti replied:
I do this once at the start of the file, assuming that err_specs is empty, for the case where the file does not contain the READ [ST]ERR commands. This gets overwritten when these commands are found.

embray commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+        )
+
+    for key, group in groupby(contents):
+        n_lines = len(list(group))
```
Here you seem to be re-parsing parts of each line; I think the regexp I provided earlier could help here; with some refactoring you could parse each line just once.

I don't understand why the `groupby` and the slicing of `lines`.  Part of that is maybe my ignorance of the file format (are comments, commands, and data always grouped together consecutively in the file)?  That's not what it looks like from the example files, so it's no clear to me why this is done this way as opposed to looping over the lines directly.

matteobachetti replied:
Basically, groupby allows to associate all the tables in the file to separate chunks, that I then read all together instead of parsing the file only once and wait for a change of data type. Since most QDP files are small (thousands of lines at most), parsing the file twice has usually no large overhead. I agree it is not the most efficient way :)

matteobachetti replied:
Ok, done with a single pass. Also, I fixed a couple of possible bugs along the way, and interpreted NOs as missing values instead of NaNs

embray commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+    contents, ncol = analyze_qdp_file(qdp_file)
+
+    with open(qdp_file) as fobj:
+        lines = list(fobj.readlines())
```
`.readlines()` returns a list to begin with on most file-like objects.

matteobachetti replied:
Oh, I thought it was an iterator. Fixed!

embray commented on astropy/io/misc/tests/test_qdp.py:
```diff
@@ -0,0 +1,117 @@
+    55045.099887 1.14467592592593e-05    -1.14467592592593e-05   0.000000        -nan
+    """
+
+    fd, path = tempfile.mkstemp()
```
you can use the tmpdir or tmp_path fixture from pytest: https://docs.pytest.org/en/stable/tmpdir.html

matteobachetti replied:
In this case, I would prefer using what the Python standard library gives me instead of a dependency that might change its behavior in the future, unless there is a compelling reason. After all, I just want a temporary file, and anything pytest can give me -as far as I understand- complicates the code instead of simplifying it

embray reviewed (changes requested):
This is helpful for understanding how this file format is interpreted and the code is very well formatted.  I think it can be simplified though.

matteobachetti commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+def is_qdp(origin, filepath, fileobj, *args, **kwargs):
+    if filepath is not None:
+        return filepath.endswith(('.qdp'))
+    return False
```
It's basically just a text file with free-format column data, with or without instructions for PLT, with or without comments. It's really difficult to distinguish it from other free-format column data :(

matteobachetti commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+        return "comment"
+
+    # Full line of NOs -> new table
+    if probe.strip("NO ") == "":
```
I'm impressed by the `_line_type_re`, but if I may, I had a hard time following that up :). 
I thought about using re's but decided otherwise because I wanted something easy to read and fix. No strong opinions.

Just one comment about the `_decimal_re`: does it understand numbers in scientific format?

matteobachetti replied:
However, I tried all doctests and this solution doesn't pass them. It fails at the very first:
```
103     Examples
104     --------
105     >>> line_type("READ SERR 3")
UNEXPECTED EXCEPTION: ValueError('Unrecognized QDP line: READ SERR 3')
```
As per the "N O N NOON OON O" line... I hope nobody is so evil to write something like that :D

matteobachetti replied:
Ok, I expanded your regular expression and now it works in all cases, including scientific notation!

matteobachetti commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+    raise ValueError(f"Unrecognized QDP line: {line}")
+
+
+def get_type_from_list_of_lines(list_of_lines):
```
Done

matteobachetti commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+
+    ncol = int(all_data_specs[0].replace("data,", "").strip())
+
+    return contents, ncol
```
Cool, done

matteobachetti commented on astropy/io/misc/qdp.py:
```diff
@@ -0,0 +1,451 @@
+
+    Returns
+    -------
+    colnames : List
```
Done

matteobachetti commented:
I don't understand what is happening in those failed tests. It looks like the logger got confused, but only in one test case :O

pllim reviewed (commented):
Probably need user facing doc too.

cc @PaulKuin

matteobachetti commented:
@embray: I think I fixed most fixable things :). I left a couple more requests for clarifications or minor "rebuttals"
@pllim: I added something to the main Read/Write docs, feel free to point out more documents to update

taldcroft reviewed (commented):
@matteobachetti - thanks, this will be a nice addition! 

One thing that is going to complicate matters is that this should really be made part of `io.ascii`. The QDP format is very much an ASCII table format so it should live in that subpackage and support the user doing `ascii.read('data.qdp', format='qdp', names=['x', 'y', 'z'])`. The machinery in `io.ascii` is a bit orthogonal to the way you've written this so far. If you just take the first step of moving it into `io.ascii` I can help with getting it to fit in.

matteobachetti commented:
Hi @taldcroft, I started moving things around. Alas, while the `io.misc` machinery is more or less straightforward, I still haven't understood how `io.ascii` works. I need some help to plug my functions in the correct places :)

taldcroft commented:
>  I need some help to plug my functions in the correct places.

Yes. The `io.ascii` machinery is very general and extensible but not very easy to understand. (I'd make it simpler if I wrote it today...). The easiest is if I just play around with it once you have done the basic changes, since I cannot from the top of my head tell you the path. 

It looks like you moved the code around. Are there other changes you are planning?

matteobachetti commented:
> It looks like you moved the code around. Are there other changes you are planning?

Not until I understand a little more :). I drafted a couple of classes based on the functionality in `basic.py`, but I need to understand some very basic stuff (For example: what delimiter should I give in `QDPsplitter` for "any number of spaces"?)

taldcroft commented:
> For example: what delimiter should I give in QDPsplitter for "any number of spaces"?

The delimiter `' '` implies any number of spaces. This is delegated to the standard library `csv.reader`.

matteobachetti commented:
@taldcroft I don't understand how to make `read` accept the `table_id` keyword

taldcroft commented:
@matteobachetti - I pushed a commit that seems to do the trick. 

One comment, can you use `names` instead of `input_colnames` for API consistency with `io.ascii`?  An analog is the `no_header` format where supplying `names` will define the column names.

matteobachetti commented:
Thanks! I'll change the keyword to `names` as requested (and I guess there will be some more API fixes like this to do)

taldcroft commented:
There are a few flake8 issues that should be cleaned up:
```
(astropy) ➜  astropy git:(add_qdp) flake8 astropy/io/ascii/qdp.py --ignore=E402,W503,F541 --max-line-length=100 
astropy/io/ascii/qdp.py:9:1: F401 'itertools.groupby' imported but unused
astropy/io/ascii/qdp.py:14:1: F401 'astropy.io.registry as io_registry' imported but unused
astropy/io/ascii/qdp.py:29:101: E501 line too long (131 > 100 characters)
astropy/io/ascii/qdp.py:443:5: F401 'astropy.io.ascii' imported but unused
```

taldcroft commented on CHANGES.rst:
```diff
@@ -177,6 +177,8 @@ astropy.io.fits
 astropy.io.misc
 ^^^^^^^^^^^^^^^
 
+- Added QDP data file handling [#11256]
```
This should be in `io.ascii` and can be a bit more verbose, perhaps "Added support for reading and writing ASCII tables in QDP (Quick and Dandy Plotter) format."

matteobachetti replied:
Done

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,552 @@
+        self.input_colnames = input_colnames
+
+    def read(self, qdp_file):
+        return read_table_qdp(qdp_file, table_id=self.table_id,
```
Another key point for API consistency is to allow the input to be a filename, a string with the data, a list of strings or a file object. This is normally done using the following:
```
        # Get a list of the lines (rows) in the table
        self.lines = self.inputter.get_lines(table, newline=newline)
```
Then the core table parsing should always expect a list of lines. I see you have support for providing the table data as a string, though this is not obvious from the docstrings.

To go 100% in the style of `io.ascii`, the standalone routines like `read_table_qdp` would not be considered part of the public API and then you would rewrite that function and all the downstream ones to accept only a list of lines.

Another option is to adapt `get_lines_from_file` to support the standard input options.

taldcroft commented on docs/io/unified.rst:
```diff
@@ -202,6 +202,7 @@ ascii.fixed_width_no_header    Yes          :class:`~astropy.io.ascii.FixedWidth
                  ascii.ipac    Yes          :class:`~astropy.io.ascii.Ipac`: IPAC format table
                 ascii.latex    Yes    .tex  :class:`~astropy.io.ascii.Latex`: LaTeX table
             ascii.no_header    Yes          :class:`~astropy.io.ascii.NoHeader`: Basic table with no headers
+                  ascii.qdp    Yes    .qdp   QDP_: Quick and Dandy Plotter files
```
Should be a `:class: ..` like the others where the class docstring has the reference info.

taldcroft replied:
Also include this new format in `astropy/docs/io/ascii/index.html`  around line 260.

matteobachetti replied:
Done

taldcroft commented on docs/io/unified.rst:
```diff
@@ -911,6 +912,15 @@ By default, ``serialize_method`` for Time columns is equal to
 
 .. _table_io_hdf5:
 
+QDP
```
It does not get a separate section here since it is "just" another ASCII format.

matteobachetti replied:
Done

matteobachetti commented:
@taldcroft next: I'm trying to supersede the write interface, but I don't understand how to do that. I commented out my definition of write so that tests pass, but at the moment the interface isn't able to write proper QDP files (with SERR/TERR instructions)

matteobachetti commented:
@taldcroft, I just noticed that you reviewed the code while I was making the last modifications. I'll work on your comments. Thanks!

taldcroft commented on astropy/io/ascii/tests/test_qdp.py:
```diff
@@ -0,0 +1,135 @@
+
+    table0 = read_table_qdp(fp.name, names=["MJD", "Rate"], table_id=0)
+    assert table0.meta["initial_comments"].lstrip().startswith("Swift")
+    assert table0.meta["comments"].lstrip().startswith("WT -- hard data")
```
Comments should be a list instead of string. See https://docs.astropy.org/en/stable/io/ascii/read.html#comments-and-metadata.

matteobachetti replied:
Done

taldcroft commented on astropy/io/ascii/tests/test_qdp.py:
```diff
@@ -0,0 +1,135 @@
+
+
+def test_roundtrip(tmpdir):
+    example_qdp = """
```
For testing it is nice to include a synthetic file with "simple" values. This makes it easier to verify that it is fully working. For example in my testing of this I used this:
```
example_qdp = """
    ! Initial comment line 1
    ! Initial comment line 2
    READ TERR 1
    READ SERR 3
    ! Table 0 comment
    !MJD            Err (pos)       Err(neg)        Rate            Error
    53000.5   0.25  -0.5   1  1.5  3.5 2
    54000.5   1.25  -1.5   2  2.5  4.5 3
    NO NO NO NO NO
    ! Table 1 comment
    !MJD            Err (pos)       Err(neg)        Rate            Error
    54000.5   2.25  -2.5   3  3.5  5.5 5
    55000.5   3.25  -3.5   4  4.5  6.5 nan
    """

dat = ascii.read(example_qdp, format='qdp', table_id=1, names=['a', 'b', 'c', 'd'])
dat

<Table length=2>
   a     a_perr  a_nerr    b       c     c_err     d   
float64 float64 float64 float64 float64 float64 float64
------- ------- ------- ------- ------- ------- -------
54000.5    2.25    -2.5     3.0     3.5     5.5     5.0
55000.5    3.25    -3.5     4.0     4.5     6.5     nan
```
You can leave this realistic Swift file if you like, but you should include something like my example and then test that every single value in both tables is expected, and that all the columns have the right name and dtype.

matteobachetti replied:
Should be done

taldcroft commented:
More changes (I made these locally for testing):
```
diff --git a/astropy/io/ascii/__init__.py b/astropy/io/ascii/__init__.py
index e10998ac1..9e1b79a19 100644
--- a/astropy/io/ascii/__init__.py
+++ b/astropy/io/ascii/__init__.py
@@ -35,6 +35,7 @@ from .latex import Latex, AASTex, latexdicts
 from .html import HTML
 from .ipac import Ipac
 from .daophot import Daophot
+from .qdp import QDP
 from .sextractor import SExtractor
 from .fixedwidth import (FixedWidth, FixedWidthNoHeader,
                          FixedWidthTwoLine, FixedWidthSplitter,
diff --git a/astropy/io/ascii/qdp.py b/astropy/io/ascii/qdp.py
index d6af72700..a62d2d6c0 100644
--- a/astropy/io/ascii/qdp.py
+++ b/astropy/io/ascii/qdp.py
@@ -536,7 +536,6 @@ class QDP(basic.Basic):
 
     """
     _format_name = 'qdp'
-    _io_registry_format_aliases = ['qdp']
     _io_registry_can_write = True
     _io_registry_suffix = '.qdp'
     _description = 'Quick and Dandy Plotter'
```
The `io_registry_format_aliases` should not be used in this case. It does require that for reading with `Table`, the `format` is given as `ascii.qdp` instead of just `qdp` but this is the preferred syntax for clarity.

matteobachetti commented on astropy/io/ascii/tests/test_qdp.py:
```diff
@@ -0,0 +1,135 @@
+    with open(path, "w") as fp:
+        print(example_qdp, file=fp)
+    with pytest.warns(AstropyUserWarning) as record:
+        table = read_table_qdp(fp.name, names=["MJD", "Rate"],
```
Right. done

matteobachetti commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,558 @@
+
+class QDPSplitter(core.DefaultSplitter):
+    """
+    Split on comma for QDP tables
```
Done

matteobachetti commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,558 @@
+        return 'comment'
+    match = _line_type_re.match(line)
+    if match is None:
+        raise ValueError(f'Unrecognized QDP line: {line}')
```
It's in the doctest above

matteobachetti commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,558 @@
+    distributed by the Swift-Gehrels mission have an extra space in one header
+    entry that makes the number of  inconsistent with the number of columns
+
+    ..
```
Done

matteobachetti commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,558 @@
+    Headers are just comments, and tables distributed by various missions
+    can differ greatly in their use of conventions. For example, light curves
+    distributed by the Swift-Gehrels mission have an extra space in one header
+    entry that makes the number of  inconsistent with the number of columns
```
Done

matteobachetti commented:
@taldcroft thanks a lot for your thorough review until now, I think I addressed all of your comments :)

matteobachetti commented:
There is a failure I don't understand. Aren't warnings allowed?

taldcroft commented:
@matteobachetti - uncaught warnings are not allowed in testing, they are converted to errors as you've seen. Using https://docs.pytest.org/en/stable/warnings.html you need to catch them and test that the expected warning got emitted.

matteobachetti commented:
But I do test it, in test_qdp.py. This error comes from the generic test in test_write.py. I guess this assumes that ascii readers don't emit warnings with default values, but in this case, this reader is made for files that can contain multiple tables, and it has to warn the user if the user does not specify which table to read.

taldcroft commented:
@matteobachetti - you can put in some special-case handling for the `qdp` format to update `kwargs` so the test passes (if indeed it fully supports writing masked data). I.e. add the appropriate `table_id` (if I have gotten things right).

matteobachetti commented:
@taldcroft I don't think the failure has to do with my PR, so whenever you want, this is ready for review.

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+    return lines
+
+
+# def _analyze_qdp_file(qdp_file):
```
Remove this commented-out function if it's not needed.

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+
+    Parameters
+    ----------
+    err_specs : dict, {'serr': [n0, n1, ...], 'terr': [n2, n3, ...]}
```
The top line should just be `err_specs : dict` (i.e. only specify the type) with the details in the next line.

See https://github.com/astropy/astropy/pull/11118/files for examples of where this is coming from.

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+
+    Other parameters
+    ----------------
+    names : list of strings
```
`list of str`  (again https://github.com/astropy/astropy/pull/11118/files).

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+    qdp_file : str
+        Input QDP file name
+
+    Other parameters
```
Docstring doesn't match function signature.

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+
+    Other parameters
+    ----------------
+    names : list of strings
```
`list of str`

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+
+    Returns
+    -------
+    tables : list of `Table` objects
```
```
list of `~astropy.table.Table`
```

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+        Number of the table to be read from the QDP file. This is useful
+        when multiple tables present in the file. By default, the first is read.
+
+    names : list of strings
```
Docstring parameters should have the same order as in the function. Also, `list of str`

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+
+    Returns
+    -------
+    tables : list of `Table` objects
```
```
list of `~astropy.table.Table`
```

taldcroft commented on astropy/io/ascii/qdp.py:
```diff
@@ -0,0 +1,578 @@
+
+
+class QDP(basic.Basic):
+    """QDP table.
```
This main docstring should be expanded to illustrate the available parameters and usage. This will be the only way people know about the `header_id`, `err_specs` etc. For a good example see the following that shows a full example of the file format and how to read. You can also include a write example. https://github.com/astropy/astropy/blob/1ed1d945aadbdb9548e37eccfcc28115e3d03ca8/astropy/io/ascii/daophot.py#L315

taldcroft reviewed (changes requested):
@matteobachetti - I took another pass and there are just docs and minor test issues to be fixed. One key thing is that you need to rebase on current master and update this to reflect the new change log system. See emails to astropy-dev (ANN: astropy plans to switch to towncrier for changelog management), in particular the one from Simon that talks about how to fix existing PRs. Hopefully the rebase will fix the CI failure as well.

Welcome to astropy development, it's never quite easy!  :smile:

matteobachetti commented:
@taldcroft should be done, including the rebase. I'm not sure the CHANGES.rst file is correct after rebase though.

pllim commented:
We don't use `CHANGES.rst ` anymore, see https://github.com/astropy/astropy/blob/main/docs/changes/README.rst

pllim commented:
@saimn , is squashing necessary here to change the changelog ?

matteobachetti commented:
The failures should be unrelated. The failing test is about a config file, and unit tests were passing up until 04db56e (the only failure was from readthedocs, bad formatting of a docstring)

pllim commented:
That is weird (the config failure). Luckily, that test is getting removed in #10877 . Unluckily, it failing is blocking the rest of the jobs for you.

saimn commented:
> @saimn , is squashing necessary here to change the changelog ?

No, just revert the changes in `CHANGES.rst` (`git checkout master -- CHANGES.rst`)

matteobachetti commented:
@pllim @taldcroft I renamed my fork's `master` branch to `main`, rebased this branch, and eliminated the changes to `CHANGES.rst` as suggested. Thanks for your help :)

matteobachetti commented:
Due to the trial-and-error path I followed and the initial proposed addition to `io.misc` then moved to `io.ascii`, there are many commits in this PR. I might give a stab at reducing the number of commits by rebasing interactively one last time, if you think it's useful (or you might just squash and merge when satisfied with the changes)

taldcroft commented:
Yes squashing the commits would be good, maybe even down to just one.

matteobachetti commented:
@taldcroft I tried to preserve the main "building blocks" of this PR, while discarding the initial io.misc interface. It's a quite big rebase, hope it works :D

taldcroft reviewed (approved):
@matteobachetti - looks great now, thanks for all the hard work and perseverance!
